File: 05--if-you-give-a-seed-a-fertilizer/a.py
#!/usr/bin/env python3
'''
Usage:
    ./a.py PATH_TO_INPUT_FILE
'''

import sys


def main():
    text = open(sys.argv[1]).read().strip()
    sections = text.split('\n\n')
    seeds_text, *map_texts = sections

    # Parse seeds
    seeds = [
        int(n)
        for n in seeds_text.split()[1:]  # Discard first item: It's "seeds:"
    ]

    # Parse maps and turn them into functions
    maps = []
    for map_text in map_texts:
        name_line, *map_lines = map_text.splitlines()
        maps.append(make_map(map_lines))

    # Solve
    locations = []
    for seed in seeds:
        x = seed
        for f in maps:
            x = f(x)
        locations.append(x)
    print(min(locations))


def make_map(lines):
    ranges = {}
    for line in lines:
        dst_start, src_start, range_length = [int(n) for n in line.split()]
        src_range = range(src_start, src_start + range_length)
        dst_range = range(dst_start, dst_start + range_length)
        ranges[src_range] = dst_range

    def f(x):
        for src_range, dst_range in ranges.items():
            if x in src_range:
                return dst_range[src_range.index(x)]
        return x

    return f


if __name__ == '__main__':
    main()

File: 05--if-you-give-a-seed-a-fertilizer/a.hs
#!/usr/bin/env runhaskell
-- Usage:
--   ./a.hs PATH_TO_INPUT_FILE

import Data.Function ((&))
import Data.List (elemIndex)
import Data.Maybe (fromJust)
import System.Environment (getArgs)

import Util (split, readInt)


data IntRange = IntRange
  { start :: Int  -- inclusive
  , end :: Int  -- exclusive
  } deriving (Show, Eq)


main = do
  [inputPath] <- getArgs
  text <- readFile inputPath
  print $ solve text


solve text =
  seeds
  & map seedToLocation
  & foldr1 min
  where
    -- Parse seeds and map data
    paragraphs = split "\n\n" text
    seedsText = head paragraphs
    mapTexts = drop 1 paragraphs

    seeds =
      seedsText
      & words
      & drop 1
      & map readInt

    mapDatas = map parseMap mapTexts

    -- Convert to functions
    functions = map toFunction mapDatas

    -- Apply all functions to a seed
    seedToLocation seed =
      foldl
        (\accValue currFn -> currFn accValue)
        seed
        functions


parseMap text =
  text
  & lines
  & drop 1
  & map (\line -> line & words & map readInt)


toFunction mapData =
  (\x ->
    case (any (\range -> rangeContains range x) srcRanges) of
      True ->
        let
          srcRange = firstBy (\range -> rangeContains range x) srcRanges
          dstRange =
            elemIndex srcRange srcRanges
            & fromJust
            & (\i -> dstRanges !! i)
          idx = x - (start srcRange)
        in
          (start dstRange) + idx
      False -> x
  )
  where
    srcRanges =
      mapData
      & map (\[_, start, length] -> IntRange start (start + length))
    dstRanges =
      mapData
      & map (\[start, _, length] -> IntRange start (start + length))


firstBy f xs =
  filter f xs
  & head


rangeContains r x =
  (start r) <= x && x < (end r)
